### ISSUE TITLE
False positive subtracting two numpy arrays

### ISSUE THREAD

**Issue by matangover** [microsoft/pyright#4003]
**Describe the bug**
An error is reported in the following valid code.
```python
from typing import Any
import numpy as np
import numpy.typing as npt

x: npt.NDArray[Any] = ...
sub1 = x - x # error
add1 = x + x # no error

y: npt.NDArray[np.int_] = ...
sub2 = y - y # no error
add2 = y + y # no error
```

**Expected behavior**
Error should not be reported.

**VS Code extension or command-line**
Pylance v2022.9.40

**Additional context**
Mypy does not report an error.

**Comment by erictraut** [comment]
This is the result of the following overload definition for the `np.ndarray` class:

```python
    @overload
    def __sub__(self: NDArray[bool_], other: _ArrayLikeBool_co) -> NoReturn: ...
```

This is the first overload among several other `__sub__` overloads. When you use `NDArray[Any]`, pyright chooses the first overload. 

Mypy's behavior for overload selection differs slightly from pyright's when `Any` is concerned. Mypy will assume that the resulting return type is `Any` in this case, but pyright chooses the first overload that matches. Neither PEP 484 nor the Python documentation clearly dictate overload behaviors, so neither of these is "wrong" or "right". We have chosen to deviate from mypy in the case because pyright is not just a type checker but also the basis for a language server, and choosing the first overload tends to produce better results for a language server. 

So I would consider this a bug in the numpy type stub. A similar bug was filed recently relating to the `cross` method, which also used a `NoReturn` return type for the first overload. I think that bug has been fixed in the most recent version of numpy. The `__sub__` overload should be fixed in the same manner.

If this is important to you, please consider filing a bug in the numpy issue tracker.

I'm going to close this bug report because I don't consider this a bug in pyright.

**Comment by matangover** [comment]
Thanks! Is there a thread open somewhere to standardize overload selection between type checkers?

**Comment by matangover** [comment]
I can't find the issue with the `cross` method, can you link to it?

**Comment by erictraut** [comment]
I'm not aware of any such thread. I think you'll find that both mypy and pyright maintainers will be convicted about our respective behaviors. We both have good justifications, so I don't know if standardization discussions will go anywhere, but if you feel strongly about it, you could certainly start a discussion.

The `cross` method issue was filed in the `pylance-release` repo. Let me know if you have trouble finding it.

**Comment by matangover** [comment]
> I think you'll find that both mypy and pyright maintainers will be convicted about our respective behaviors.

Of course, and rightly so! But since I've seen your efforts in standardizing other aspects of the type system, I assume you also appreciate the value of reaching an agreement. Especially for the benefit of the users.

**Comment by matangover** [comment]
I found this issue is apparently already fixed in numpy main https://github.com/numpy/numpy/pull/22359

**Comment by erictraut** [comment]
Standardization is useful in some areas and less so in others. For example, I'd push back pretty strongly if someone were to propose standardizing inference or type narrowing rules. I think overload resolution is in the gray area. I think that both mypy and pyright have principled reasons for adopting their current behaviors, and forcing one or both to change in the name of standardization may not be the best answer for their respective user bases. But I'm open to having that discussion if you want to pursue it.

**Comment by matangover** [comment]
Just read what you wrote in https://github.com/microsoft/pyright/issues/2521 about the various options for how to handle overload selection, and the reasoning behind the current choice. It was very interesting. I was wondering about one thing you said:
> If I were to rank the above options based on type checking alone, I'd say that [...] option 3 is by far the worst.

Why so? Is it not 'type-checking correct' to say that the return type of an ambiguous call is a union of all possible overloads' return types?

> You might argue that we should have different behaviors when type checking is disabled, but we're very reluctant to do this [...]

Definitely agree that enabling a different behavior only when type checking is enabled is confusing. What do you think of the following options -

1. Adding an independent option for strict overload checking? FWIW, I'd love to enable it...
2. Adding a visual indication in the inferred type (in the hover tooltip and reveal_type) showing that the type was selected arbitrarily. E.g. something like `Ambiguous[float]` which behaves exactly like a `float`.
3. Adding a diagnostic (perhaps disabled by default, and enabled in strict mode) which says 'Ambiguous call to overloaded function due to unknown argument type. The first overload was selected.'

**Comment by matangover** [comment]
Thinking about it, option 3 above (diagnostic) would probably solve my concern the best, and is probably also easiest to implement (?). What bothers me is that currently the arbitrary overload selection happens transparently and pyright basically says "yes, I know the type of this variable is XXX" while in fact it doesn't know that. If it would inform me that it's making an informed guess, I'd be much happier.

**Comment by erictraut** [comment]
> Why so? Is it not 'type-checking correct' to say that the return type of an ambiguous call is a union of all possible overloads' return types?

Returning the union of all types will almost always result in false positives. There have been some discussion about introducing a "weak union", a non-type-safe version that can be treated like `Any` for type checking purposes but like a union for language server purposes (for completion suggestions, for example), but this discussion didn't go very far.

> Adding an independent option for strict overload checking

We're unlikely to do this. We have a general principle that strictness checks should not change the evaluated type. Strictness checks should only control which diagnostics are reported. (I introduced `strictListInference`, `strictSetInference` and `strictDictInference ` before establishing this principle, and I've regretted it since.)

> Adding a visual indication in the inferred type

I don't think this makes sense. Hover text is associated with an identifier, not a call expression.

> Adding a diagnostic (perhaps disabled by default, and enabled in strict mode) which says 'Ambiguous call to overloaded function due to unknown argument type.

Maybe, although this would require significant replumbing of our overload logic, which is already very complex. It would also have significant performance impact because pyright's analyzer wouldn't be able to stop after the first matching overload.

As I said above, I've thought pretty deeply about this problem, and I'm convicted that the current behavior in pyright is defensible and consistent with other design principles, so I'm not inclined to make a change to the behavior — at least not at this time.

**Comment by matangover** [comment]
> Returning the union of all types will almost always result in false positives

Ah. I understand

> It would also have significant performance impact because pyright's analyzer wouldn't be able to stop after the first matching overload

Hmm, I meant that pyright would behave exactly as it does now - it would select the first matching overload - and at that moment it would also emit the additional diagnostic. Is that difficult to implement?

**Comment by matangover** [comment]
Oh, I understand. Pyright would need to always continue checking if another overload matches the call, before emitting that diagnostic.
